aptos-move/aptos-vm/src/verifier/event_validation.rs

```diff
@@ -0,0 +1,184 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::move_vm_ext::SessionExt;
+use aptos_framework::RuntimeModuleMetadataV1;
+use move_binary_format::{
+    access::{ModuleAccess, ScriptAccess},
+    errors::{Location, PartialVMError, VMError, VMResult},
+    file_format::{
+        Bytecode, CompiledScript,
+        SignatureToken::{Struct, StructInstantiation},
+    },
+    CompiledModule,
+};
+use move_core_types::{
+    account_address::AccountAddress, language_storage::ModuleId, vm_status::StatusCode,
+};
+use std::collections::HashSet;
+
+const EVENT_MODULE_NAME: &str = "event";
+const EVENT_EMIT_FUNCTION_NAME: &str = "emit";
+
+fn metadata_validation_err(msg: &str) -> Result<(), VMError> {
+    Err(metadata_validation_error(msg))
+}
+
+fn metadata_validation_error(msg: &str) -> VMError {
+    PartialVMError::new(StatusCode::EVENT_METADATA_VALIDATION_ERROR)
+        .with_message(format!("metadata and code bundle mismatch: {}", msg))
+        .finish(Location::Undefined)
+}
+
+/// Validate event metadata on modules one by one:
+/// * Extract the event metadata
+/// * Verify all changes are compatible upgrades (existing event attributes cannot be removed)
+pub(crate) fn validate_module_events(
+    session: &mut SessionExt,
+    modules: &[CompiledModule],
+) -> VMResult<()> {
+    for module in modules {
+        let mut new_event_structs =
+            if let Some(metadata) = aptos_framework::get_metadata_from_compiled_module(module) {
+                extract_event_metadata(&metadata)?
+            } else {
+                HashSet::new()
+            };
+
+        // Check all the emit calls have the correct struct with event attribute.
+        validate_emit_calls(&new_event_structs, module)?;
+
+        let original_event_structs =
+            extract_event_metadata_from_module(session, &module.self_id())?;
+
+        for member in original_event_structs {
+            // Fail if we see a removal of an event attribute.
+            if !new_event_structs.remove(&member) {
+                metadata_validation_err("Invalid change in event attributes")?;
+            }
+        }
```

Comment:
@wrwg it's a bit hacky that we have to implement the compatibility check logic here in a slightly different way. Really should offer a way to extend the logic of the original backward compatibility checker.